File: Sources/CodeRim/Utilities/TokenFormatter.swift
import Foundation

enum TokenNumberStyle: String, CaseIterable, Codable, Sendable {
    case compact
    case detailed
}

struct TokenFormatter: Sendable {
    func string(from value: Int64, style: TokenNumberStyle, locale: Locale = .current) -> String {
        switch style {
        case .detailed:
            return value.formatted(.number.locale(locale).grouping(.automatic).precision(.fractionLength(0)))
        case .compact:
            return compactString(from: value, locale: locale)
        }
    }

    private func compactString(from value: Int64, locale: Locale) -> String {
        let magnitude: (divisor: Double, suffix: String)?
        switch value {
        case 999_500_000...:
            magnitude = (1_000_000_000, "B")
        case 999_500...:
            magnitude = (1_000_000, "M")
        case 1_000...:
            magnitude = (1_000, "K")
        default:
            magnitude = nil
        }

        guard let magnitude else {
            return String(value)
        }

        let scaled = Double(value) / magnitude.divisor
        let fractionDigits = scaled >= 100 ? 0 : (scaled >= 10 ? 1 : 2)
        let number = scaled.formatted(
            .number
                .locale(locale)
                .grouping(.never)
                .precision(.fractionLength(0...fractionDigits))
        )
        return number + magnitude.suffix
    }
}

File: Windows/src/CodeRim.Core/Services/TokenFormatter.cs
using System.Globalization;

namespace CodeRim.Core.Services;

public enum TokenNumberStyle
{
    Compact,
    Detailed
}

public static class TokenFormatter
{
    public static string Format(long value, TokenNumberStyle style) =>
        style == TokenNumberStyle.Detailed
            ? value.ToString("N0", CultureInfo.CurrentCulture)
            : Compact(value);

    private static string Compact(long value)
    {
        var (divisor, suffix) = value switch
        {
            >= 999_500_000 => (1_000_000_000d, "B"),
            >= 999_500 => (1_000_000d, "M"),
            >= 1_000 => (1_000d, "K"),
            _ => (0d, string.Empty)
        };

        if (divisor == 0)
        {
            return value.ToString(CultureInfo.CurrentCulture);
        }

        var scaled = value / divisor;
        var decimals = scaled >= 100 ? 0 : scaled >= 10 ? 1 : 2;
        var format = decimals == 0 ? "0" : $"0.{new string('#', decimals)}";
        return scaled.ToString(format, CultureInfo.CurrentCulture) + suffix;
    }
}
